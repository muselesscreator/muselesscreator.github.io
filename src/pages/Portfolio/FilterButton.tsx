import classNames from 'classnames';

type FilterButtonProps = {
  isActive: boolean;
  filter: string;
  handleFilterChange: (category: string) => () => void;
};
const FilterButton = ({
  isActive,
  filter,
  handleFilterChange,
}: FilterButtonProps) => {
  return (
    <li className="filter-item">
      <button
        className={classNames({ active: isActive })}
        onClick={handleFilterChange(filter)}
      >
        {filter}
      </button>
    </li>
  );
};
export default FilterButton;
