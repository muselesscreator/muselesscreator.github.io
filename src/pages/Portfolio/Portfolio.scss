.project-item {
  .project-links {
    svg {
      transform: translate(0, 2px);
      display: inline-block;
    }
    a {
      margin-left: 5px;
    }
  }
}
