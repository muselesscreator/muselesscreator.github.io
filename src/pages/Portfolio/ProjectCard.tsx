import { FaEye } from 'react-icons/fa';

import { StrictDict } from '@muselesscreator/strict-dict';

import { ProjectData } from './types';


export const classNames = StrictDict({
  img: 'project-img',
  iconBox: 'project-item-icon-box',
  title: 'project-title',
  category: 'project-category',
}) as Record<string, string>;

const ProjectCard = ({
  title,
  description,
  category,
  image,
  projectUrl,
}: ProjectData) => (
  <a href={projectUrl} target="_blank" rel="noopener noreferrer">
    <figure className={classNames.img}>
      <div className={classNames.iconBox}>
        <p>{description}</p>
        <FaEye />
      </div>
      <img src={image} alt={title} loading="lazy" />
    </figure>
    <h4 className={classNames.title}>{title}</h4>
    <p className={classNames.category}>{category}</p>
  </a>
);

export default ProjectCard;

