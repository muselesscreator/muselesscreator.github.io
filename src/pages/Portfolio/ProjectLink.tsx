type ProjectLinkProps = {
  IconComponent: React.ComponentType,
  href?: string,
  label?: string,
};

const ProjectLink = ({
  IconComponent,
  href,
  label,
}: ProjectLinkProps) => (
  (href && label) && (
    <p>
      <IconComponent />: 
      <a href={href} target="_blank" className="project-link" rel="noopener noreferrer">
        {label}
      </a>
    </p>
  )
);

export default ProjectLink;

