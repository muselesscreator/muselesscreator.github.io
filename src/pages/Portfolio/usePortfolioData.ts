import React from 'react';

import projects, { ProjectCategories } from './data';

const usePortfolioData = () => {
  const [projectFilter, setProjectFilter] = React.useState(ProjectCategories.all);
  const handleCategoryChange = (value: unknown) => () => setProjectFilter(value as string);
  const handleFilterChange = (value: unknown) => setProjectFilter(value as string);
  const filteredProjects = projects.filter(
    (project: any) => project.category === projectFilter || projectFilter === ProjectCategories.all
  );
  return {
    filteredProjects,
    handleCategoryChange,
    handleFilterChange,
    projectFilter,
  };
};

export default usePortfolioData;
