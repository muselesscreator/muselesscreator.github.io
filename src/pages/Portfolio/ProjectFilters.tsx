import Select from '~/components/Select';
import { StrictDict } from '@muselesscreator/strict-dict';

import FilterButton from './FilterButton';
import { ProjectCategories } from './data';

type ProjectFiltersProps = {
  projectFilter: string;
  handleCategoryChange: (category: string) => () => void;
  handleFilterChange: (value: unknown) => void;
};

export const filterOptions = [
  { label: ProjectCategories.all, value: ProjectCategories.all },
  { label: ProjectCategories.public, value: ProjectCategories.public },
  { label: ProjectCategories.openSource, value: ProjectCategories.openSource },
];

export const categories = [
  ProjectCategories.all,
  ProjectCategories.public,
  ProjectCategories.openSource,
];

export const classNames = StrictDict({
  list: 'filter-list',
  selectBox: 'filter-select-box',
}) as Record<string, string>;

const ProjectFilters = ({
  projectFilter,
  handleCategoryChange,
  handleFilterChange,
}: ProjectFiltersProps) => {
  return (
    <div>
      <ul className={classNames.list}>
        {categories.map((category) => (
          <FilterButton
            key={category}
            isActive={projectFilter === category}
            handleFilterChange={handleCategoryChange}
            filter={category}
          />
        ))}
      </ul>
      <div className={classNames.selectBox}>
        <Select
          label="Select category"
          options={filterOptions}
          value={projectFilter}
          handleChange={handleFilterChange}
        />
      </div>
    </div>
  );
};

export default ProjectFilters;
