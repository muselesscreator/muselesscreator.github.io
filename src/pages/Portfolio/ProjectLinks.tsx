import { RiNpmjsFill, RiGithubFill } from 'react-icons/ri';

import ProjectLink from './ProjectLink';
import { ProjectData } from './types';

const ProjectLinks = ({
    githubUrl,
    githubRepo,
    packageName,
    packageUrl,
}: ProjectData) => (
  <div className="project-links">
    <ProjectLink IconComponent={RiGithubFill} href={githubUrl} label={githubRepo} />
    <ProjectLink IconComponent={RiNpmjsFill} href={packageUrl} label={packageName} />
  </div>
);

export default ProjectLinks;

