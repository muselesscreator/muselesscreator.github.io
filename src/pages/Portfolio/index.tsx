import ProjectCard from './ProjectCard';
import ProjectLinks from './ProjectLinks';
import usePortfolioData from './usePortfolioData';
import ProjectFilters from './ProjectFilters';
import './Portfolio.scss';

const Portfolio = () => {
  const {
    projectFilter,
    handleCategoryChange,
    handleFilterChange,
    filteredProjects,
  } = usePortfolioData();
  return (
    <article className="portfolio active">
      <header>
        <h2 className="h2 article-title">Portfolio</h2>
      </header>
      <section className="projects">
        <ProjectFilters
          projectFilter={projectFilter}
          handleCategoryChange={handleCategoryChange}
          handleFilterChange={handleFilterChange}
        />
        <ul className="project-list">
          {filteredProjects.map((project) => (
            <li className="project-item active" key={project.title}>
              <ProjectCard {...project} />
              <ProjectLinks {...project} />
            </li>
          ))}
        </ul>
      </section>
    </article>
  );
};

export default Portfolio;
