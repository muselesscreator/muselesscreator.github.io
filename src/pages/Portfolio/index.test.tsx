import {
  beforeEach,
  vi,
  it,
  test,
  describe,
  expect,
  Mock,
} from 'vitest';
import { shallow, setMockName } from '@muselesscreator/react-shallow-snapshot';

import usePortfolioData from './usePortfolioData';
import Portfolio from '.';

vi.mock('./ProjectCard', () => ({ default: 'ProjectCard' }));
vi.mock('./ProjectLinks', () => ({ default: 'ProjectLinks' }));
vi.mock('./ProjectFilters', () => ({ default: 'ProjectFilters' }));
vi.mock('./usePortfolioData', () => ({ default: vi.fn() }));

const hookProps = {
  projectFilter: 'test-project-filter',
  handleCategoryChange: vi.fn(),
  handleFilterChange: vi.fn(),
  filteredProjects: [
    { title: 'project1-title', type: 'project1-type', value: 'project1-value' },
    { title: 'project2-title', type: 'project2-type', value: 'project2-value' },
    { title: 'project3-title', type: 'project3-type', value: 'project3-value' },
  ],
};
setMockName(hookProps.handleCategoryChange, 'hookProps.handleCategoryChange');
setMockName(hookProps.handleFilterChange, 'hookProps.handleFilterChange');

(usePortfolioData as Mock).mockReturnValue(hookProps);

let el: any;
describe('Portfolio page', () => {
  beforeEach(() => {
    el = shallow(<Portfolio  />);
  });
  test('snapshot', () => {
    expect(el.snapshot).toMatchSnapshot();
  });
  it('displays ProjectFilters', () => {
    const filtersEl = el.instance.findByType('ProjectFilters')[0];
    expect(filtersEl.props).toEqual({
      projectFilter: hookProps.projectFilter,
      handleCategoryChange: hookProps.handleCategoryChange,
      handleFilterChange: hookProps.handleFilterChange,
    });
  });
  it('displays filtered projects with cards and links', () => {
    const items = el.instance.findByClassName('project-item active');
    expect(items).toHaveLength(hookProps.filteredProjects.length);
    items.forEach((item: any, index: number) => {
      const card = item.findByType('ProjectCard')[0];
      expect(card.props).toEqual(hookProps.filteredProjects[index]);
      const links = item.findByType('ProjectLinks')[0];
      expect(links.props).toEqual(hookProps.filteredProjects[index]);
    });
  });
});
