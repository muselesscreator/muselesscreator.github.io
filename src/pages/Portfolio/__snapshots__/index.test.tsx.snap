// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Portfolio page > snapshot 1`] = `
"<article
  className="portfolio active"
>
  <header>
    <h2
      className="h2 article-title"
    >
      Portfolio
    </h2>
  </header>
  <section
    className="projects"
  >
    <ProjectFilters
      handleCategoryChange={[Function hookProps.handleCategoryChange]}
      handleFilterChange={[Function hookProps.handleFilterChange]}
      projectFilter="test-project-filter"
    />
    <ul
      className="project-list"
    >
      <li
        className="project-item active"
      >
        <ProjectCard
          title="project1-title"
          type="project1-type"
          value="project1-value"
        />
        <ProjectLinks
          title="project1-title"
          type="project1-type"
          value="project1-value"
        />
      </li>
      <li
        className="project-item active"
      >
        <ProjectCard
          title="project2-title"
          type="project2-type"
          value="project2-value"
        />
        <ProjectLinks
          title="project2-title"
          type="project2-type"
          value="project2-value"
        />
      </li>
      <li
        className="project-item active"
      >
        <ProjectCard
          title="project3-title"
          type="project3-type"
          value="project3-value"
        />
        <ProjectLinks
          title="project3-title"
          type="project3-type"
          value="project3-value"
        />
      </li>
    </ul>
  </section>
</article>"
`;
